state/protocol/badger/snapshot.go

```diff
@@ -486,7 +488,17 @@ func (q *EpochQuery) Current() protocol.Epoch {
 		return invalid.NewEpochf("could not get current EpochCommit (id=%x) for block %x: %w", status.CurrentEpoch.CommitID, q.snap.blockID, err)
 	}
 
-	epoch, err := inmem.NewCommittedEpoch(setup, commit)
+	firstHeight, _, epochStarted, _, err := q.retrieveEpochHeightBounds(setup.Counter)
+	if err != nil {
+		return invalid.NewEpochf("could not get current epoch height bounds: %s", err.Error())
```

Comment:
My guess why you used `err.Error()` instead of just wrapping the error struct itself:
* If this call ever errors, we know something is badly broken. Therefore, we don't want to propagate the error itself, as it might coincide with some benign error cases.

If that is correct, could you please document this explicitly in the code? 

```suggestion
		// Any errors are symptoms of potential state corruption, bugs, or uncovered edge cases and therefore
		// safety-critical exceptions. We strip error type info to avoid any confusion with benign sentinels.  
		return invalid.NewEpochf("could not get current epoch height bounds: %s", err.Error())
```

Reply:
> Therefore, we don't want to propagate the error itself, as it might coincide with some benign error cases.

The reasoning is that, if a function returns an error which is unexpected (ie. not a documented possible error return), that error should also not have a type which might be considered benign further up the stack. In a sense, there is an extra error type representing an exception which should never resolve to any documented error type via `errors.Is`. We want to cut off the error chain at the point an error is deemed to be an unexpected exception, which can be accomplished by converting the error to a string. 

Adding the comments like you suggested works, but I feel like to be strict, this pattern should be applied pretty commonly, and actually creating a utility function/type to cut off the error chain in these situations would be more practical. For example:
```
err := doSomething()
if err != nil {
  if errors.Is(err, Sentinel) {
    return fmt.Errorf("failed to xyz: %w", err)
  } 
  // rather than `fmt.Errorf(...)` which might preserve sentinel error information, use a utility which always
  // cuts off the error chain, but preserves the text context.
  return util.Exception(err)
}
```


### Example
Here's an example of the kind of situation I am thinking of, where not cutting off the error chain can be harmful. It is difficult to know while working on a function in isolation whether it might cause a situation like this.

```go

var ErrNotFound = errors.New("not found")

type store map[string]string

// Error returns:
//  - ErrNotFound
func (s store) retrieve(key string) (string, error) {
	val, ok := s[key]
	if !ok {
		return "", ErrNotFound
	}
	return val, nil
}

// No Errors expected
func b(s store) error {
	// b stores then retrieves a value. Since the value must always exist,
	// no errors from retrieve are expected.
	s["1234"] = "5678"
	_, err := s.retrieve("1234")
	if err != nil {
		return fmt.Errorf("unexpected error: %w", err)
	}
	return nil
}

// a calls two subfunctions:
//  - retrieve, which might return an expected error.
//  - b, which can only return an unexpected error.
//
// Error returns:
//  - ErrNotFound
func a(s store) error {
	// a retrieves some value - we aren't sure whether the value exists in the store
	// at this point, so ErrNotFound is an expected error return
	val, err := s.retrieve("asdf")
	if err != nil {
		return fmt.Errorf("could not get asdf key: %w", err)
	}

	err = b(s)
	if err != nil {
		// b has no expected errors
		// when b has an unexpected error, it has concrete type ErrNotFound
		// if b does not strip the type information from its errors before returning
		// then the caller (a) might return an error which is:
		//  - semantically unexpected
		//  - AND resolves as an expected error type via errors.Is(...)
		return fmt.Errorf("unexpected error: %w", err)
	}

	_ = val
	return nil
}
```

Reply:
I had exact same problem, because both `errors.Is and errors.As` scan the whole chain of errors. I have dealt using the same approach as Jordan. I feel like we should pair documentation and runtime guarantees. 
I think it's quite beneficial to have a utility function which transforms error into a generic one which can't be mapped to other sentinels. If some function doesn't return sentinels it's better to make sure that it won't in all cases. This will allow us to detect issues earlier rather than masking them.

Reply:
Something like this:
```goalng
// Exception wraps an error, which can be potentially a sentinel error.
// IT does NOT IMPLEMENT an UNWRAP method, so the enclosed error's type
// cannot be accessed. Therefore, methods such as `errors.As` and `errors.Is`
// do not detect the Exception as a known sentinel error.  
type Exception struct {
	err error 
}

func (e Exception) Error() string { return e.err.Error() }

func StripSentinelTypes(e error) error {
	return Exception{e}
}
```

Reply:
Created a PR adding this https://github.com/onflow/flow-go/pull/3999